Consolidate network namespace handling

bleggett opened the pull request:
Fixes: #27449

This is the first step towards fixing #27449 and #27450 - Removing dependence on `ip netns` from the internal netns lib, and consolidating netns creation/handling into one lib/spot, while not requiring mount pins for every netns creation, the way we currently do - most things (incl. tests) rarely need this.

This PR removes a few netns libs from the repo:

- `containernetworking/plugins/pkg/ns`
- `vishvananda/netns` from all but one spot (`pkg/datapath/linux/devices_controller.go`, where a sister lib wants a typecast)

in favor of an internal lib `"github.com/cilium/cilium/pkg/netns"` which avoids pinning netnses, and only has a dep on the `unix` package.

General goals of the `github.com/cilium/cilium/pkg/netns` package:

- To be used in all Cilium code that needs to create new netnses.
- Supports creating unpinned netnses owned by Cilium code.
- Supports running closures in netnses previously pinned _outside_ of Cilium code, by other processes (needed for CNI)
- No support for pinning netnses owned by Cilium code, as we never actually need to do that.
- Stick to stdlib `sys/unix`

Looking for:

- Feedback on `pkg/netns` API surface/impl/general approach

Generally the intent is to make sure that code that _creates_ netnses is solely responsible for closing them.

Code that is running out-of-process from the process that is the netns owner sometimes does need to run some closure in the netns context, but should not be burdened with netns lifecycle concerns, or be responsible for closing/cleanup, in virtually all cases. The API surface is designed to discourage the current overreliance on named-pin-based disposal/cleanup.

cc @ti-mo to make sure I interpreted his intent correctly.

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
+	Name   string
+}
+
+type NetNsHandle interface {
```
Are there going to be multiple implementations of a netns? Is there a need for an interface?

bleggett replied:
Nope! I was originally thinking I might need to mock the OS primitives for tests but just using real ones and running privileged is what I've gone with, so will remove this.

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
 
-// RemoveIfFromNetNSIfExists removes the given interface from the given
-// network namespace.
+const bindMountPath = "/run/netns" /* Bind mount path for named netns */
```
I think the canonical path is still `/var/run/netns` according to the `ip netns` man pages: https://man7.org/linux/man-pages/man8/ip-netns.8.html.

I'd need to audit our code a bit closer, but aside from the health endpoint, I don't think we interact with entries in /var/run/netns at all, and in the cilium-health case, it's just to start the process there. There's no real need for the netns to be pinned, it only creates more file system state to deal with during restarts etc.

bleggett replied:
Yes, this is something to hash out.

`vishvananda/netns` uses `/run/netns` for it's pins.

`ip netns xxx` uses `/var/run/netns`

I don't think it actually matters which location is used. Latter would make these discoverable by `ip netns` tooling, which is probably desirable.

bleggett replied:
(for my benefit/further clarification) so the actual _correct_ location to bind netnses is any named subdirectory under `/run` - `/var/run` is a legacy location that is now (as of like 2011) always symlinked to `/run`.

`/run/netns` -> this is the path that `ip netns` uses, by convention.

`/run/{docker, containerd, whatever}/netns` -> this is the path style other things use - netnses bound here are still fully valid and usable, they just won't show up by default in tools like `ip netns` that hardcode the search path to `/run/netns` - this is intentional.

I don't think we need pins at all as previously discussed but if we do, it might be cleaner to put ours under `/run/cilium/netns`

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
+
+type NetNsHandle interface {
+	// Returns a file descriptor representing this object's network namespace
+	Fd() int
```
```suggestion
	FD() int
```

(https://go.dev/talks/2014/names.slide#5: `Acronyms should be all capitals, as in ServeHTTP and IDProcessor.`)

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
+}
+
+type NetNsHandle interface {
+	// Returns a file descriptor representing this object's network namespace
```
Not just any file descriptor; this returns NetNSHandle's underlying file descriptor.

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
+//
+// If you need a pinned network namespace, call `CreateNewNetns` with pin == true, or call Pin() on
+// the returned handle.
+func CreateNewNetns(netNSName string, pin bool) (NetNsHandle, error) {
```
I'd just call this `New()`. On the calling side, this reads as `netns.New()`; `netns.CreateNewNetns()` stutters.

See my other comment around the design of Pin(). Netns should not have implicit names. Sure, they can be accessed by their pin path, but those are more like symlinks than real names.

Also, no need for the pin argument. Bool args are difficult to parse mentally. A New() should just be followed by a Pin() if that's what the caller intends to do.

bleggett replied:
Yep agree, was already feeling the bool arg should go in favor of an explicit Pin by caller.

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
+const bindMountPath = "/run/netns" /* Bind mount path for named netns */
+
+type netNSHandle struct {
+	handle int
```
`handle` doesn't convey what this is, this should be called `fd`.

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
+	// Returns a file descriptor representing this object's network namespace
+	Fd() int
+
+	// Cleans up this instance of the network namespace; if this instance
```
I tend to avoid language like 'clean up' because it typically implies removing some kind of file system state, but it's generally rather imprecise and handwavy. Let's be specific, `Close the network namespace. If this handle is the last reference, the namespace will be destroyed.`

Close() should never unpin, that would make it impossible for a netns handle to outlive the current process. Correct resource management in Go would mean a netNSHandle needs a runtime finalizer that calls Close() on the underlying fd if the handle is garbage collected, otherwise we run the risk of leaking file descriptors if someone forgets to call Close().

bleggett replied:
Yep, as with `New` agree they should be separate. 

To your point tho, I do not know yet if we actually *need* to pin across process anywhere - except for health endpoint, where we do retain the FD in the creating process until after the secondary process terminates, and do explicitly clean it up.

If there are explicitly no places in the code where a pinned netns needs to outlive the creating process, it may be cleaner to do this, even if it makes this less useful as a generic netns library.

ti-mo replied:
> where we do retain the FD in the creating process until after the secondary process terminates, and do explicitly clean it up

We need to clean up because `ip netns add` creates file system state. (`ip` is a separate process in and of itself, so it kinda has to)  If we'd simply launch the process in an unpinned netns, we wouldn't need to clean anything up to begin with. :slightly_smiling_face: And if the Pod dies, so does the health listener.

There may be folks still running Cilium outside of containers today, so as long as we keep the pidfile management code in the health endpoint launcher, we can always kill the old process before starting a new one, and the netns will die along with it.

bleggett replied:
Yep, I was going to actually try launching the subproc unpinned, if that works agree we may not need pins at all (assuming I don't find other weird usages in the rest of the code)

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
+	// instance is currently Pin-ed.
+	Close() error
+
+	// "Pins" or bind-mounts the netns created here to a specific path in nsfs.
```
I'd avoid documenting implementation details like bind mounting, especially since this is an interface definition.

I'm not convinced we need pinning to begin with. iirc cilium-health is the only consumer, and the current code allocates an endpoint, creates a new netns from scratch and kill the old process, so it may as well live in an anonymous netns. On top of that, it spawns this process within the agent container, so its lifecycle is completely bound to the agent pod. IMO this would be better off as just a goroutine pinned to a thread living in another netns, but we're not going to touch that during this work.

In case we do have a need for pinning, I'd like to design this API exactly like the pinning APIs in ebpf-go, e.g. See https://github.com/cilium/ebpf/blob/5ab77468412f8b5a8dd4d5541a5bd59558c765cb/map.go#L1297. Pin() should take the

bleggett replied:
_If_ (big IF) we do need pins, it might make sense to be opinionated about where pinned NSes created by us live, since I doubt there's any need for us to manage or handle netnses we don't create, and we would probably prefer to discourage callers from trying to do that.

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
+
+	// Tests equality of two network namespace handles, typically by checking device and inode of the
+	// underlying FDs
+	Equal(other NetNsHandle) bool
```
This seems to be only used from within the netns' own tests, no need to export this.

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
+	Name   string
+}
+
+type NetNsHandle interface {
```
Note: here and everywhere else: `s/NetNs/NetNS/` and `s/Netns/NetNS/` please.

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
 //
-// Example usage of this function:
+// That's what this is for.
+func DoByPinName(netNSName string, runInside func(NetNsHandle) error) error {
```
The new `netns` package should have two main exported functions: `New()` (as outlined previously) and `Open()`. to open a netns from a path passed to the CNI binary by the container runtime. I'm assuming these are paths like `/proc/123/ns/net`, I doubt container runtimes temporarily pin things to /var/run/netns. The code this replaced (`ns.GetNS()`) just calls `os.Open()` on the given path, so this doesn't seem right.

bleggett replied:
Yep. I was trying make it so calling code doesn't have to concern itself with where exactly netnses are pinned in the FS but when actually looking at usages, it's pretty clear only the CNI plugin does this anyway and it's already a path.

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
+
+	// Do runs the provided func within a new thread+netns, without changing the current thread's netns
+	//
+	// Note that since network namespaces are set *per thread*, both the current thread and the new thread will
```
> both the current thread and the new thread

Why does the current thread need to be locked?

bleggett replied:
It doesn't, comment is vestigial.

ti-mo commented on pkg/netns/netns.go:
```diff
@@ -4,88 +4,302 @@
+	//
+	// The provided closure should _not_ call runtime.UnlockOSThread() at all,
+	// or else undefined behavior should be expected.
+	Do(runInside func(NetNsHandle) error) error
```
Why expose NetNsHandle to the callee? The caller already has a NetNsHandle, since they're already calling `Do()` on it. This should just be `Do(f func() error) error`.

bleggett replied:
This may not be needed.

We have some existing code where we do things with e.g. interfaces within the Do closure and need a contract that the FD of the netns is available inside the closure.

For now in this WIP, I am just doing it this way so moving that code is straightforward.

ti-mo commented on cilium-dbg/cmd/cleanup.go:
```diff
@@ -516,7 +516,7 @@ func removeRoutesAndLinks(routes map[int]netlink.Route, links map[int]netlink.Li
 
 func removeNamedNetNSs(netNSs []string) error {
```
I think this is already dead code, `ciliumCleanup.netNSs` is never assigned to, feel free to remove it.

bleggett replied:
On line 153 we are actually scanning all the pinned netnses and looking for pins with a given prefix, to clean up any extant pins.

I don't think we need this at all,  but again if we do we could just keep the pins in `/run/cilium/netns` and nuke that whole dir on cleanup instead.

ti-mo commented on pkg/datapath/linux/node_linux_test.go:
```diff
@@ -2147,14 +2146,14 @@ func (s *linuxPrivilegedIPv6OnlyTestSuite) TestArpPingHandlingForMultiDevice(c *
 	c.Assert(err, check.IsNil)
 
-	netns0, err := netns.ReplaceNetNSWithName("test-arping-netns0")
+	netns0, err := netns.CreateNewNetns("test-arping-netns0", false)
```
Note: see my comment on the interface definition, but none of these need names or pin paths.

bleggett replied:
Agree. Touching these tests made that clear.

ti-mo reviewed (commented):
Thanks for picking this up! I've renamed the PR to something more descriptive and left a few preliminary comments.

derailed commented on cilium-health/launch/endpoint.go:
```diff
@@ -87,28 +85,30 @@ func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing
+	return healthNetNS.Do(func(*netns.NetNSHandle) error {
+		out, err := exec.Command(prog, args...).CombinedOutput()
+		if err == nil && len(out) > 0 {
+			log.Warn(out)
```
I think we should be more specific here. Perhaps reusing log with fields so users can see the command?

derailed commented on cilium-health/launch/endpoint.go:
```diff
@@ -291,13 +285,20 @@ func LaunchAsEndpoint(baseCtx context.Context,
+	// Run the health binary inside a netnamespace. Since `Do()` implicitly does `runtime.LockOSThread`
+	// the exec'd binary is guaranteed to inherit the corret netnamespace.
+	if doRunErr := netNS.Do(func(*netns.NetNSHandle) error {
+		if err := cmd.Run(); err != nil {
```
we should just return cmd.Run()

derailed commented on cilium-health/launch/endpoint.go:
```diff
@@ -320,8 +321,8 @@ func LaunchAsEndpoint(baseCtx context.Context,
-	if err = configureHealthRouting(info.ContainerName, epIfaceName, node.GetNodeAddressing(), mtuConfig); err != nil {
-		return nil, fmt.Errorf("Error while configuring routes: %s", err)
+	if err = configureHealthRouting(&netNS, epIfaceName, node.GetNodeAddressing(), mtuConfig); err != nil {
+		return nil, fmt.Errorf("Error while configuring routes for containername %s : %s", info.ContainerName, err)
```
Prefer %q just in case...

derailed commented on cilium-health/launch/endpoint.go:
```diff
@@ -267,11 +262,10 @@ func LaunchAsEndpoint(baseCtx context.Context,
+	netNS, err := netns.New()
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("Error while creating endpoint netns: %s", err)
```
We should use %w

derailed commented on cilium-health/launch/endpoint.go:
```diff
@@ -281,7 +275,7 @@ func LaunchAsEndpoint(baseCtx context.Context,
 
-		if err = netlink.LinkSetNsFd(epLink, int(netNS.Fd())); err != nil {
+		if err = netlink.LinkSetNsFd(epLink, int(netNS.FD())); err != nil {
 			return nil, fmt.Errorf("failed to move device %q to health namespace: %s", epIfaceName, err)
```
Use %w

derailed commented on pkg/datapath/connector/veth.go:
```diff
@@ -7,19 +7,19 @@ import (
+	return netns.DoByPinPath(netNsPinPath, func(_ *netns.NetNSHandle) error {
 		err := link.Rename(srcIfName, dstIfName)
 		if err != nil {
 			return fmt.Errorf("failed to rename veth from %q to %q: %s", srcIfName, dstIfName, err)
```
Prefer %w

derailed commented on pkg/netns/netns.go:
```diff
@@ -4,106 +4,165 @@
+	fd, err := netns.New()
 	if err != nil {
-		return nil, fmt.Errorf("Unable to open netns %s: %s", path, err)
+		innerError = fmt.Errorf("Unable to create new netns: %s", err)
```
Why aren't we returning here?

bleggett replied:
leftover from old approach, fixed.

derailed commented on pkg/netns/netns.go:
```diff
@@ -4,106 +4,165 @@
-		return nil, fmt.Errorf("Unable to open netns %s: %s", path, err)
+		innerError = fmt.Errorf("Unable to create new netns: %s", err)
 	}
+	fmt.Printf("inner FD: %d", fd)
```
Left over debugging??

derailed commented on pkg/netns/netns.go:
```diff
@@ -4,106 +4,165 @@
+		innerError = fmt.Errorf("Unable to create new netns: %s", err)
 	}
+	fmt.Printf("inner FD: %d", fd)
+	newNetns = NetNSHandle{fd: int(fd)}
```
We might as well just return at this point vs creating a new var?

derailed commented on pkg/netns/netns.go:
```diff
@@ -4,106 +4,165 @@
+func DoByPinPath(netNSPath string, runInside func(*NetNSHandle) error) error {
+	fd, err := unix.Open(netNSPath, unix.O_RDONLY|unix.O_CLOEXEC, 0)
+	// We aren't keeping this and we don't own the pin or the netns it refers to
+	defer unix.Close(fd)
```
Do we need to check if we have an error prior to closing in the `defer`? 
Also unix.Close returns an error.

bleggett replied:
In practice it doesn't hurt but yeah check before defer, ty!

bleggett replied:
As far as not checking `unix.Close` , that was to avoid using defer closures everywhere just for error reporting - if we fail to close a unix FD there is not really anything the caller can do about it.

if we would prefer to log it for debugging I can add.

derailed commented on plugins/cilium-cni/cmd/cmd.go:
```diff
@@ -540,11 +542,16 @@ func (cmd *Cmd) Add(args *skel.CmdArgs) (err error) {
+			// but get an FD to the remote one, without moving this thread inside it
+			fd, err := unix.Open(args.Netns, unix.O_RDONLY|unix.O_CLOEXEC, 0)
+			// We aren't keeping this and we don't own the pin or the netns it refers to
+			defer unix.Close(fd)
```
nit: no error check on close

bleggett replied:
As far as not checking `unix.Close` , that was to avoid using defer closures everywhere just for error reporting - if we fail to close a unix FD there is not really anything the caller can do about it.

if we would prefer to log it for debugging I can add.

derailed reviewed (changes requested):
@farcaller Thank for this update!

derailed commented on pkg/datapath/linux/devices_test.go:
```diff
@@ -28,7 +28,7 @@ import (
 
 type DevicesSuite struct {
-	currentNetNS                      netns.NsHandle
+	// currentNetNS                      netns.NsHandle
```
nit: Clean up?

bleggett replied:
Yep, ty, fixed!

derailed commented on pkg/datapath/linux/devices_test.go:
```diff
@@ -57,7 +57,7 @@ func (s *DevicesSuite) SetUpSuite(c *C) {
 	s.prevConfigEnableIPv6NDP = option.Config.EnableIPv6NDP
 	s.prevConfigIPv6MCastDevice = option.Config.IPv6MCastDevice
-	s.currentNetNS, err = netns.Get()
+	// s.currentNetNS, err = netns.Get()
```
nit: Clean up?

derailed reviewed (changes requested):
@bleggett Thank you for the updates!

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+//
+// Cilium code should not pin or bind-mount network namespaces it creates, this function is
+// purely for interacting with network namespaces created and managed outside of Cilium.
+func DoByPinPath(netNSPath string, runInside func(*NetNSHandle) error) error {
```
Instead of this open-and-do function, I prefer separate `Open(path string) NetNS, error` and subsequent `netns.Do()` calls. Allowing building blocks to be composed is nice and allows reusing the `*NetNS`.

bleggett replied:
In general, aside from the CNI which _must_ handle pinned NSes created elsewhere, the goal here was to keep this as a one-off to discourage people in the Cilium codebase from leveraging this API on external pinned NS paths, and to discourage them from reusing NetNSes created by pin path, specifically.

It is more composable to split them, at the cost of depending on the caller to explicitly close to avoid a FD leak.

But given the other calls are now also making that assumption (and that we're OK deferring to the finalizer as a safety net), agree it makes sense to split.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	"golang.org/x/sys/unix"
+)
+
+type NetNSHandle struct {
```
Let's just call this NetNS, the `Handle` doesn't add much value.

derailed reviewed (approved):
@bleggett Thank you for the updates!

ti-mo commented on cilium-health/launch/endpoint.go:
```diff
@@ -87,28 +85,30 @@ func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing
+	prog := "bash"
+	args := []string{"-c"}
 	routeCmds := []string{}
 	for _, rt := range routes {
```
Instead of this contraption, could we move the `for _, rt := range routes {` into the `healthNetNS.Do` and just execute individual `exec.Command`s directly instead of rendering them into a single `bash -c`? That should be more robust and less code.

bleggett replied:
Yep, I recall briefly wondering why this wasn't done before. I can't see any reason not to.

bleggett replied:
I've improved it a bit,  in that we exec individually and I ditched some of the munging ~~It's still wrapped in `bash -c` because `ToIPCommand()` doesn't really return a proper `Command` style format and I don't wanna get too far afield with changes, but could drop that later.~~ and dropped bash

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	fd int
+}
+
+// CreateNewNetns creates a network namespace and
```
Please pick a column width and stick to it.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	if err != nil {
+		return NetNSHandle{fd: -1}, fmt.Errorf("Unable to create new netns: %w", err)
+	}
+	return NetNSHandle{fd: int(fd)}, innerError
```
Returning a bare NetNS like this runs the risk of runaway resource leaks if the caller forgets to Close(). I'd add a `newNetNS(int fd) NetNS` that calls `runtime.SetFinalizer(ns, NetNS.Close())` before returning the new NetNS.

Nit: variable `fd` is already int.

bleggett replied:
Yep, I tried to mitigate that originally with API design that didn't require callers to own and close the NetNS handle.

Now that they do, `SetFinalizer` will be something of a safety net, tho not a perfect solution.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+//
+// This means the provided closure should _not_ call runtime.UnlockOSThread() at all,
+// or else undefined behavior should be expected.
+func (h *NetNSHandle) Do(runInside func(*NetNSHandle) error) error {
```
```suggestion
func (h *NetNSHandle) Do(f func() error) error {
```

As far as I can see, the `*NetNSHandle` parameter isn't used anywhere. I think I've mentioned this in a previous comment, but I don't think we should ever pass the handle. Close() seems a little pointless to call from within the Do() closure, and recursive calls to Do() are absolutely not safe.

A secondary benefit to removing the NetNSHandle argument from function passed to Do() is that we should then be able to unexport `NetNSHandle` by renaming it to `netNS`. Callers should never synthesize their own netns handles, as they wouldn't contain a finalizer.

Also nit: functions passed as arguments are commonly just called `f`.

bleggett replied:
> As far as I can see, the *NetNSHandle parameter isn't used anywhere. I think I've mentioned this in a previous comment, but I don't think we should ever pass the handle. 

Yes, you did, and thanks for reminding me again - I twiddled the existing code in `plugins/cilium-cni/cmd/cmd.go` that led me to include it in the first place, so we don't need it at all anymore, will remove.

> recursive calls to Do() are absolutely not safe.

AFAIK it should be fine in the sense that we would as a result have multiple threads in the same netns and aside from eventually exhausting the thread pool, that's fine in a Linux/go context, but I might be missing something.

Either way, will drop it since it's not needed.

> Also nit: functions passed as arguments are commonly just called f.

Yes, it's a common convention I'm not fond of since like all one-char arg names it's not terribly descriptive, but will fix it :D

bleggett replied:
Fixed this, missing that it was totally unused now was a bonehead move on my part, apologies for making you repeat yourself.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	}()
+
+	localHandle := NetNSHandle{
+		fd: int(fd),
```
Variable `fd` is already int.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+// the returned handle.
+func New() (NetNSHandle, error) {
+	var innerError error
+	var newNetns NetNSHandle
```
Here and everywhere across the codebase: please settle on a single variable naming convention for all NetNSHandles. This makes greps and refactoring easier. I'd call it `ns` wherever possible.

bleggett replied:
I'll do a pass in this PR and try to mention the convention somewhere.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	var newNetns NetNSHandle
+	restore, err := saveRestoreCurrentThreadNetns()
+	if err != nil {
+		return newNetns, err
```
We should return `nil, err` here.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+//
+// If you need a pinned network namespace, call `CreateNewNetns` with pin == true, or call Pin() on
+// the returned handle.
+func New() (NetNSHandle, error) {
```
We should always return `*NetNSHandle` so the handle isn't copied by accident. This is important as a copy won't have a finalizer.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	}
+	defer restore()
+
+	// this invokes the UNSHARE syscall, which creates a new netns *and moves the current thread*
```
Document this on `createAndMoveToNetNetNS` instead.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+}
+
+// FD returns the underlying file descriptor of the NetNSHandle
+// Note that FDs should not be used for comparing or testing equality, for that
```
`equals` is unexported and only used in tests. Comparing NetNSHandles would be rather uncommon for a caller, so I'd drop this comment.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	return nil
+}
+
+// Do runs the provided func within a new thread+netns, without changing the current thread's netns
```
```suggestion
// Do runs the provided func in the netns without changing the calling thread's netns.
```

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+
+// Do runs the provided func within a new thread+netns, without changing the current thread's netns
+//
+// Note that since network namespaces are set *per thread*, the thread the closure is running inside
```
I'd replace the two following paragraphs with:
```
// The code in f and any code called by f must NOT call [runtime.LockOSThread],
// as this will leave the goroutine created by Do permanently pinned to an OS thread.
```

bleggett replied:
I'll change `will` to `could` since it kinda depends on whether `f` terminates or calls `UnlockOSThread`, but yea, will use this.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+
+// NOTE If we ever need to test equality of netNSes outside of test code,
+// make this public.
+func (h *NetNSHandle) equal(other *NetNSHandle) bool {
```
I'd implement this free function/test helper inside the test suite for now, this doesn't need to be implemented on the type itself. It's unlikely to ever be useful anywhere else.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+// this function was called, and unlock the current OS thread.
+func saveRestoreCurrentThreadNetns() (func(), error) {
+	runtime.LockOSThread()
+	hostNSFD, err := getCurrent()
```
I think we want to use the term 'orig' to denote the original netns. Just call this `orig`.

bleggett replied:
I was trying to convey that this will restore your thread to whatever the current netns was for that thread at the time this was called. `orig()` is a little opaque.

ti-mo replied:
I mean `orig, err := getCurrent()`

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+		// Creating a new netns with UNSHARE also moves the current thread to that new netns,
+		// which is not always what the caller wants.
+		// So lock current thread and save a handle to current network namespace now so we can switch back.
+		defer runtime.UnlockOSThread()
```
Why use these `defer`s here? If the goal is to explicitly ignore return values, use `_ =`. Also makes the code look wrong at first glance since these will be executed in reverse order.

bleggett replied:
Yep strictly unnecessary here, ty.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+		// So lock current thread and save a handle to current network namespace now so we can switch back.
+		defer runtime.UnlockOSThread()
+		// We aren't keeping this and we don't own the pin or the netns it refers to
+		// so defer closure.
```
Outdated comment?

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	if err != nil {
+		// flaking out early, so unlock current thread
+		runtime.UnlockOSThread()
+		return nil, fmt.Errorf("Failed to open current namespace: %v", err)
```
Nit: errors imply failure. Also, avoid caps at the start of error strings, they're hard to parse when the error gets wrapped. Also, `%w` is the verb for wrapping errors.
```suggestion
		return nil, fmt.Errorf("get current namespace: %w", err)
```

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	}
+
+	return func() {
+		// Creating a new netns with UNSHARE also moves the current thread to that new netns,
```
Outdated comment?

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+
+// new creates a new network namespace, sets it as current and returns
+// a handle to it.
+func createAndMoveToNewNetNS() (int, error) {
```
Let's use familiar terminology and terse function names:
```suggestion
func unshare() (int, error) {
```

bleggett replied:
Fair enough - `unshare` probably isn't familiar or intuitive terminology to most people tho.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	}, nil
+}
+
+// new creates a new network namespace, sets it as current and returns
```
Outdated comment?

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+}
+
+// getFromPath gets a file descriptor to a network namespace
+// identified by the path
```
```suggestion
// pinned at path.
```

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	if err := unix.Unshare(unix.CLONE_NEWNET); err != nil {
+		return -1, err
+	}
+	return getCurrent()
```
`getCurrent()` is fallible, and the caller can't distinguish between a failure of the unix.Unshare() above and the subsequent `getCurrent()`. If `getCurrent()` fails for whatever reason, the caller needs to restore the current thread to the original netns.

I'd lift this call to `getCurrent()` back to the caller and leave `unshare()` as a convenience wrapper.

bleggett replied:
> getCurrent() is fallible

Strictly speaking yea, tho unlikely (`/proc` unreadable). Will fix ty.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	// Start the callback in a new green thread so that if we later fail
```
Let's use Go terminology here: `Start the function in a new goroutine ...`. This comment also doesn't accurately reflect what it does or why it's effective. I think the original comment you removed was better:
```
// The current goroutine is locked to an OS thread and we've failed
// to undo state modifications to the thread. Returning without unlocking
// the goroutine will make sure the underlying OS thread dies.
```

I think maybe linking to the documentation of `runtime.UnlockOSThread` would be useful here as well. We didn't do that in the `haveManagedNeighbor` probe, since it was a one-off.

https://pkg.go.dev/runtime#UnlockOSThread:
`If the caller made any permanent changes to the state of the thread [...] it should not call this function and thus leave the goroutine locked to the OS thread until the goroutine (and hence the thread) exits.`

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	return (s1.Dev == s2.Dev) && (s1.Ino == s2.Ino)
+}
+
+// saveRestoreCurrentThreadNetns returns a cleanup func (which callers should `defer`)
```
This only documents what the cleanup function does and is redundant with the second paragraph. I'd call this reset/revert instead of cleanup, which sort of implies removing files or other resources.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+//
+// the cleanup func will move the thread back to the netns it was running under when
+// this function was called, and unlock the current OS thread.
+func saveRestoreCurrentThreadNetns() (func(), error) {
```
The closure returned by this function attempts to undo the effects of `Unshare()`, but the parent function only calls LockOSThread(). That doesn't make any sense.

I'd either name this `lockAndEnter(fd int) (func(), error)` and make this function enter the netns after locking the OS thread, or call it `lock() (func(), error)` and have it only deal with `runtime.(Unlock)OSThread()` and a separate `enter(fd int) (func(), error)` that deals with `Unshare/Setns`.

bleggett replied:
Yep, it's too tightly coupled to being called in concert with other helper funcs, fair enough.

bleggett replied:
> The closure returned by this function attempts to undo the effects of Unshare(), but the parent function only calls LockOSThread(). That doesn't make any sense.

It's designed to unconditionally restore the current thread to whatever netns it had when you first invoked `saveRestoreCurrentThreadNetns`. It doesn't care what the netns was when you called that.

If you call it without `unshare` and then invoke the returned func, then it will no-op, which is fine.

It doesn't undo `unshare`, it unconditionally moves the current thread back to whatever netns it had when the function was first invoked, and doesn't care about `unshare` - if the netns hasn't changed, then it's a no-op.

bleggett replied:
> all it lock() (func(), error) and have it only deal with runtime.(Unlock)OSThread() and a separate enter(fd int) (func(), error) that deals with Unshare/Setns.

The latter would always need to be fenced with the former to be deterministic, so it doesn't make much sense to me to separate them here, IMO - it just makes all usages more complex and footgunny.

thoughts?

ti-mo replied:
To me, what makes the most sense is a `lockAndEnter()` that locks the goroutine and unshares, then returns a function that sets the thread back to orig and unlocks the goroutine.

bleggett replied:
> To me, what makes the most sense is a lockAndEnter() that locks the goroutine and unshares, then returns a function that sets the thread back to orig and unlocks the goroutine.


What's here does all of that, except the unshare-ing/moving namespaces is factored out of the save/restore function, because it's not strictly relevant to save/restore and might be done differently depending on whether we are `unshare`-ing, or opening an existing netns pin path (as we still must, at least, still, for CNI).

Could pass the pin path to `saveRestoreCurrentThreadNS` but that makes the private helper func a little more complex and less generalized, and doesn't change the outward-facing behavior at all.

I might be missing a benefit tho?

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,200 @@
+	// to switch the namespace back to the original one, we can safely
+	// leave the thread locked to die without a risk of the current thread
+	// left lingering with incorrect namespace.
+	var innerError error
```
How about using a `sync.ErrGroup` instead? That avoids the awkward error variable scoping issues.

bleggett replied:
Oh that's new to me. Much better.

ti-mo reviewed (changes requested):
I focused more on the core API implementation this round, but one nit I'd like to see addressed in the codebase overall: let's use `ns, err := netns.New()` wherever possible.

This makes all occurrences familiar and removes mental overhead when reading, in addition to the `grep` genefits, e.g. to quickly check if all `netns.New()` have a matching `ns.Close()`. This is hard to do if the vars have random names.

ti-mo commented on pkg/datapath/linux/devices_controller_test.go:
```diff
@@ -244,8 +231,13 @@ func TestDevicesController(t *testing.T) {
 
-	withFreshNetNS(t, func(ns netns.NsHandle) {
-
+	newNet, err := netns.New()
```
Tip: create a new file `pkg/testutils/netns.go` with:

```go
// SPDX-License-Identifier: Apache-2.0
// Copyright Authors of Cilium

package testutils

import (
	"testing"

	"github.com/cilium/cilium/pkg/netns"
)

// NewNetNS returns a new network namespace.
func NewNetNS(tb testing.TB) *netns.NetNSHandle {
	tb.Helper()

	ns, err := netns.New()
	if err != nil {
		tb.Fatal(err)
	}
	tb.Cleanup(func() {
		ns.Close()
	})

	return &ns
}
```

and simply `ns := testutils.NewNetNS(t)` eveywhere.

bleggett replied:
Yep, better next step after removing most/all of the local test helper funcs.

bleggett replied:
though this is somewhat incompatible with making `NetNSHandle` private.

squeed commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,214 @@
+			return fmt.Errorf("Unable to create new netns: %w", err)
+		}
+		innerFD, err = getCurrent()
+		fmt.Printf("woo %d", innerFD)
```
debug snuck in :-)

bleggett replied:
UGH ty

squeed commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,214 @@
+// under when this function was called, and unlock the current OS thread.
+//
+// If the netns of the current thread has not changed, this is effectively a no-op.
+func saveRestoreCurrentThreadNetNS() (func(), error) {
```
There is a much simpler way to do this; don't call UnlockOSThread.  As per the standard librarly, if a goroutine terminates without a balanced number of UnlockOSThread() calls, the posix thread will also be terminated and cleanup is not required.

So, you an instead write a decorator like this:

```go
func runInThread(fn func() error) error {
    wg := sync.WaitGroup{}
    wg.Add(1)
    var err error
    go func() {
        runtime.LockOSThread()
        err = fn()
    }
    wg.Done()
    return err
}
```

and there's no need to worry about missing a revert or changing thread properties.

bleggett replied:
Yes, I considered that, and it will work, but I was concerned that locking and burning threads and forcing the runtime to terminate and clean them as the only mechanism might create more churn and work for the runtime than doing best-effort unlock first, and then letting the thread die as a last resort only if that fails.

So while that will work, using it as the default rather than the failsafe _might_ be more expensive. I do not have enough knowledge of the Go runtime to know for sure, but intuitively it feels true.

I'm not against this tho.

Thoughts?

ti-mo replied:
Even though we're still only talking about _dozens_ of OS threads, letting them die unconditionally feels heavy-handed. We know the pitfalls and how to implement this correctly without letting threads die, so let's do that on the happy path and go for thread termination when we run into an error setting the thread back to the original netns.

chaunceyjiang commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
```
Perhaps you can add a `//go:build linux`, even though the filename has a `_linux` suffix.

bleggett replied:
They do the same thing, and only one or the other is required, not both - I see the suffix-approach used elsewhere so I assume that is the preferred convention in this codebase, but happy to be corrected.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -168,19 +168,19 @@ func (h *NetNS) Do(f func() error) error {
 
 	return func() {
-		_ = unix.Setns(hostNSFD, unix.CLONE_NEWNET)
+		_ = unix.Setns(orig, unix.CLONE_NEWNET)
```
This is a crucial mistake. We can't ignore this error. If setting the thread back to its original netns fails, we should never call UnlockOSThread, as this will release the thread back into the schedulable pool.

bleggett replied:
Entirely correct, ugh. I wish I could reasonably fence this with unit tests. Well, at least I'm failing in public :D

ti-mo commented on cilium-health/launch/endpoint.go:
```diff
@@ -281,23 +271,29 @@ func LaunchAsEndpoint(baseCtx context.Context,
+	log.Debugf("Spawning health endpoint with command %q %q within its own netNS", binaryName, args)
+
+	// Run the health binary inside a netnamespace. Since `Do()` implicitly does `runtime.LockOSThread`
+	// the exec'd binary is guaranteed to inherit the corret netnamespace.
```
```suggestion
	// the exec'd binary is guaranteed to inherit the correct netnamespace.
```

ti-mo commented on pkg/datapath/linux/devices_controller_test.go:
```diff
@@ -244,8 +230,9 @@ func TestDevicesController(t *testing.T) {
-	withFreshNetNS(t, func(ns netns.NsHandle) {
-
+	ns := netns.NewNetNS(t)
+	// Run tests inside their own temp netns to avoid pollution
```
All of these comments are a little bit superfluous, I think this speaks for itself.

ti-mo commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -638,41 +640,48 @@ type netlinkFuncs struct {
-func makeNetlinkFuncs(netns netns.NsHandle) (*netlinkFuncs, error) {
-	h, err := netlink.NewHandleAt(netns)
+// Will use the current netns for scoping, call this func within its own net namespace if desired.
+func makeNetlinkFuncs() (*netlinkFuncs, error) {
```
cc @joamaki 

The `netns` argument has been removed from this function because the way it was used previously (basically only `netns.None()`) was a bit redundant. `netlink.NewHandle` already dials a new netlink socket with netns pid -1, so passing `netns.None()` is implicit behaviour.

@bleggett I suggest we go one step further and remove the call to `netlinkNetNS.Get()` since it's functionally equivalent to omitting the Namespace field in `netlink.*Options`. This allows us to drop the direct dependency on `vishvananda/netns`.

All callers to `makeNetlinkFuncs()` already seem to live within a `ns.Do()`, and I would prefer if we keep doing this explicitly instead of passing a netns handle as an argument. Right now, this requires converting the output of `ns.FD()` to a `vishvananda/netlink.NsHandle`, which is a little awkward to begin with.

bleggett replied:
will take a look, it is awkward and I'm not happy about it.

bleggett replied:
Ah ok, yeah it looks like it's an implicit assumption in the netlink lib that `-1` is the same as "use the current netns".

I checked this before when I noticed that we were "just" getting a handle to the current netns, but missed that if you provide nothing it assumes that anyway. Dropped.

ti-mo commented on pkg/testutils/netns/netns.go:
Does this need to be in its own package? I find calling `testutils.NewNetNS()` clearer. `netns.NewNetNS()` stutters and we already use the `netns` name for the base package.

bleggett replied:
Unfortunately it does, to avoid a `require` dependency loop with `testutils` itself. Otherwise I agree with you, and that's how I did it at first.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,213 @@
+	fd int
+}
+
+// CreateNewNetns creates a network namespace and returns a handle to it.
```
Outdated comment.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,213 @@
+// new network namespace will be returned.
+//
+// The namespace created by this call is NOT pinned, or bind-mounted, and so will be
+// closed/lost when the process terminates.
```
```suggestion
// closed/lost when the last process in the namespace terminates, and when the resulting handle
// is either Close()d or garbage collected.
```

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,213 @@
+		if err != nil {
+			return fmt.Errorf("Unable to obtain and lock current netns: %w", err)
+		}
+		defer restore()
```
I think we should only call `restoreUnlock()` when `unshare()` and `getCurrent()` are succesful, and let the OS thread die otherwise, so, at the end of the closure, do:

```go
if err := restoreUnlock(); err != nil {
    return fmt.Errorf("restoring and unlocking OS thread: %w", err)
}
return nil
```

bleggett replied:
Yeah I guess since we can rely on the runtime to reap the thread if we don't unlock it, we have an "out" if the defer somehow doesn't run, so they're safe to drop.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,213 @@
+	// Force this into a dedicated + threadlocked goroutine to avoid disturbing the caller's
+	// thread.
+	g.Go(func() error {
+		restore, err := saveRestoreCurrentThreadNetNS()
```
This name still confuses me. How about: `restoreUnlock, err := lockOSThreadNetNS()`. That conveys both the meaning of the callee and the function it returns.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,213 @@
+// purely for interacting with network namespaces created and managed outside of Cilium.
+//
+// Callers should still explicitly Close their NetNS when done.
+func NewFromPin(netNSPath string) (*NetNS, error) {
```
Maybe this?

```suggestion
func OpenPinned(netNSPath string) (*NetNS, error) {
```

Crucial difference; we're not creating a (new) namespace here, we're opening an fd to an existing one.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,213 @@
+	// will be invoked, or when. Just that the go runtime will try to invoke it
+	// at some point after the `ns` instance goes out of scope, if the caller
+	// forgets to explicitly Close()
+	runtime.SetFinalizer(ns, func(f *NetNS) {
```
I think this technically works, since it checks all the boxes in the SetFinalizer documentation, but this runs the risk of someone coming in later and closing over another variable in this function. We can pass a function pointer, like so:

```suggestion
	runtime.SetFinalizer(ns, (*NetNS).Close)
```

A method's receiver is implemented as the function's first argument under the hood.

bleggett replied:
Ah, cheeky, ty.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,213 @@
+	return ns, nil
+}
+
+const SO_NETNS_COOKIE = 71
```
This was recently added to x/sys/unix, I don't think we need this anymore.

bleggett replied:
Ah indeed it was, ty

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,213 @@
+}
+
+// unshare creates a new network namespace, sets it as the current thread's
+// net namespace and returns a handle to it.
```
This no longer returns a handle.

ti-mo commented on pkg/testutils/netns/netns.go:
```diff
@@ -0,0 +1,29 @@
+	if err != nil {
+		tb.Fatal(err)
+	}
+	if ns == nil {
```
Can this happen?

bleggett replied:
No, this is excessive paranoia and might have been copypasta'd by me from a previous netns helper func in the tests. Checking `err` is enough.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,213 @@
+			return err
+		}
+		// unconditionally restore the snapshotted netns on termination, no matter what
+		defer restore()
```
Could we remove the use of `defer` here? It makes it slightly easier to reason about and verify we're doing things in the right order:

```go
restoreUnlock, err := lockOSThreadNetNS()
...
unix.Setns()
ferr := f()
if err := restoreUnlock(); err != nil {
    return err
}
return ferr
```

We also need to return an error from `restoreUnlock` and make sure to never swallow it. If something goes wrong here, this must be passed back to the caller and logged.

bleggett replied:
Yep, removed, as per https://github.com/cilium/cilium/pull/29993#discussion_r1471754515

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,213 @@
+		return nil, fmt.Errorf("get current namespace: %w", err)
+	}
+
+	return func() {
```
```suggestion
	return func() error {
```
Let's make this function fallible so we can catch and log all errors.

ti-mo commented on pkg/netns/netns_linux.go:
```diff
@@ -0,0 +1,213 @@
+	g.Go(func() error {
+		restore, err := saveRestoreCurrentThreadNetNS()
+		if err != nil {
+			return fmt.Errorf("Unable to obtain and lock current netns: %w", err)
```
Lowercase error strings, please. I thought we had a linter for this. :thinking:

bleggett replied:
golangci-lint definitely has a linter for this.

ti-mo commented on pkg/datapath/linux/routing/routing_test.go:
```diff
@@ -297,3 +285,8 @@ func linkExistsWithMAC(c *C, macAddr mac.MAC) bool {
 	return false
 }
+
+func runFuncInNetNS(c *C, run func() error) {
```
Could this be removed in favor of `testutils.NewNetNS`/`ns.Do()`?

bleggett replied:
Yes, and I considered it, but in general trying to mimic what the test cases already did to avoid making this change too big.

This I will just redo tho, it's 2 refs.

bleggett replied:
...I forgot that `migrate_test` uses it too, and that's why I left it. Still, not so many refs, will drop.

ti-mo commented on pkg/datapath/connector/veth.go:
```diff
@@ -7,22 +7,26 @@ import (
+	ns, err := netns.NewFromPin(netNsPinPath)
+	if err != nil {
+		return err
+	}
```
Missing `defer ns.Close()`.

bleggett replied:
Agh.

ti-mo reviewed (changes requested):
Thanks for your continued efforts! I think this is getting close to where it needs to be, but there are a few mistakes that need to be ironed out. If everything in this round gets addressed, I think we should be good for merge.

pippolo84 reviewed (approved):
LGTM for my codeowners

ti-mo commented on pkg/datapath/linux/route/route_linux_test.go:
```diff
@@ -466,9 +466,12 @@ func runListRules(t *testing.T, family int, fakeIP, fakeIP2 *net.IPNet) {
+	ns := netns.NewNetNS(t)
+	ns.Do(func() error {
+		for _, tt := range tests {
+			t.Run(tt.name, func(t *testing.T) {
```
`t.Run()` spawns goroutines, so these subtests now all run in the host namespace. By lifting `defaultRules, _ := ListRules(family, nil)` out of the `setupWant` above, you solidified the assumption that these would all run from the hostns. I pushed a patch that reverted this.

ti-mo reviewed (approved):
Just did a final pass, I'll approve and push up the last changes I made locally. A few things in cni needed fixing.

ti-mo committed:
link: ignore missing interfaces in DeleteByName

This allows surfacing unexpected errors, but ignores interfaces that
are already absent.

Signed-off-by: Timo Beckers <timo@isovalent.com>

tommyp1ckles commented on cilium-health/launch/endpoint.go:
```diff
@@ -82,70 +79,69 @@ func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing
-		log.WithField("netns", netns).WithField("command", cmd).Debug("Adding route")
-		routeCmds = append(routeCmds, cmd)
+		cmd := rt.ToIPCommand(dev)
+		prog := cmd[0]
```
I assume this slice size is implied to be >=2 from the ToIPCommand function, but it may be worth adding a bounds check in case that assumption ever changes.

squeed commented on cilium-dbg/cmd/post_uninstall_cleanup.go:
```diff
@@ -514,19 +504,6 @@ func removeRoutesAndLinks(routes map[int]netlink.Route, links map[int]netlink.Li
 	return nil
 }
 
-func removeNamedNetNSs(netNSs []string) error {
```
Why is this being removed? I agree that this is probably good to remove, but removing it should probably be its own commit, just for clarity's sake.

bleggett replied:
This removed the pinned netns by name - we are not pinning this anymore so we do not need to remove the pin/do cleanup.

squeed commented on cilium-health/launch/endpoint.go:
```diff
@@ -320,7 +318,15 @@ func LaunchAsEndpoint(baseCtx context.Context,
-	if err = configureHealthRouting(info.ContainerName, epIfaceName, node.GetNodeAddressing(), mtuConfig); err != nil {
+	routes, err := getHealthRoutes(node.GetNodeAddressing(), mtuConfig)
+	if err != nil {
+		return nil, fmt.Errorf("Error while configuring routes for containername %q: %s", info.ContainerName, err)
```
nit: `getHealthRoutes()` doesn't configure, it just calculates, right? This log line is a bit misleading if so.

squeed commented on plugins/cilium-cni/cmd/cmd.go:
```diff
@@ -468,20 +469,22 @@ func (cmd *Cmd) Add(args *skel.CmdArgs) (err error) {
 	for _, epConf := range configs {
-		if err = netns.RemoveIfFromNetNSIfExists(netNs, epConf.IfName()); err != nil {
+		if err = ns.Do(func() error {
+			return link.DeleteByName(epConf.IfName())
```
What happens if this link doesn't exist -- will this return an error?

bleggett replied:
`DeleteByName` returns nil if the link doesn't exist, which is the same as the old function.

squeed reviewed (dismissed):
Looks basically good; I'd like to see it split in to a few more commits. One for the new API, one to move all test code, one for trivial changes, non-trivial changes in individual commits, and lastly one to remove old code. It would make reviewing a lot easier. Thanks!

tommyp1ckles commented on cilium-health/launch/endpoint.go:
```diff
@@ -82,70 +79,69 @@ func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing
+		prog := cmd[0]
+		args := cmd[1:]
+		log.Debugf("Running \"%s %+v\"", prog, args)
+		out, err := exec.Command(prog, args...).CombinedOutput()
```
Would we want to restrict warning on this to just stderr?

bleggett replied:
Maybe - I left it like it was previously tho.

ti-mo committed:
cilium-dbg: remove netns cleanup code

This was already redundant when running Cilium inside a container, since
the nsfs instance inside the container is bound to the container's
lifecycle. Running Cilium outside of a container is currently rather
involved and not officially supported. Remove the netns cleanup code.

When the container exits, cilium-health and its enclosing namespace also
disappears.

Follow-up commits will remove the code in package netns being called here.

Signed-off-by: Timo Beckers <timo@isovalent.com>

bleggett committed:
netns: clean up API, reimplement in pure Go without dependencies

The previous netns package had a few problems. It shelled out to iproute2,
and it depended on both containernetworking/plugins/pkg/ns and vishvananda/netns,
which lead to some idiosyncratic API.

This commit addresses these issues and takes care of some much-needed API cleanup:

- Create a new netns with New()
- Open an existing pinned netns with OpenPinned()
- Execute code within the netns with ns.Do()
- Close with Close()

Pinning network namespaces is not supported, as there is currently little reason
for doing so. In case the requirement pops up again later, it can always be added.

All tests now use anonymous (non-pinned) network namespaces, and the netns created
for cilium-health also no longer leaves an entry in /var/run/netns.

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>
Co-authored-by: Timo Beckers <timo@isovalent.com>

tommyp1ckles reviewed (approved):
ipam changes lgtm

sayboras reviewed (approved):
LGTM from vendor team
